### PR TITLE
additional benchmarks: utfcpp

### PR DESCRIPTION
**Pull request by bkmgit**
- utfcpp release 3.2.2 https://github.com/nemtrif/utfcpp

### CODE REVIEW

**Comment by bkmgit** [comment]
@lemire Basic infrastructure added to enable comparisons with utfcpp and cuneicode.  Additional tests can be added for cuneicode, currently only a conversion from utf32 to utf8 is done which is based on the example provided in the cuneicode repository.  Can add additional tests, but feedback on additions would be great before progressing further.

**Comment by bkmgit** [comment]
@lemire Basic infrastructure added to enable comparisons with utfcpp and cuneicode.  Additional tests can be added for cuneicode, currently only a conversion from utf32 to utf8 is done which is based on the example provided in the cuneicode repository.  Can add additional tests, but feedback on additions would be great before progressing further.

**Comment by lemire** [comment]
The Cuneicode import is sizeable (11 MB and over a 1000 files). I would make it an optional target (we can have CMake download a dependency on demand). I don't want all our users to grab a copy of Cuneicode.

Could you limit this PR to [utfcpp](https://github.com/nemtrif/utfcpp)? To follow our convention so far, the subdirectory should be `utfcpp` (and then we can bump the versions as needed).

**Comment by bkmgit** [comment]
Ok. Renamed it. Updating the code a little for it to compile correctly.

**Comment by lemire** [comment]
Looks reasonable !!! Let us run the tests.

**Comment by bkmgit** [comment]
Will debug:
```
D:\a\simdutf\simdutf\benchmarks\src\benchmark.cpp(2077,30): error C2672: 'utf8::utf8to16': no matching overloaded function found [D:\a\simdutf\simdutf\build\benchmarks\src\simdutf_benchmarks_benchmark.vcxproj]
  D:\a\simdutf\simdutf\benchmarks\src\benchmark.cpp(2110,30): error C2672: 'utf8::utf16to8': no matching overloaded function found [D:\a\simdutf\simdutf\build\benchmarks\src\simdutf_benchmarks_benchmark.vcxproj]
  D:\a\simdutf\simdutf\benchmarks\src\benchmark.cpp(2131,29): error C2672: 'utf8::utf8to32': no matching overloaded function found [D:\a\simdutf\simdutf\build\benchmarks\src\simdutf_benchmarks_benchmark.vcxproj]
  D:\a\simdutf\simdutf\benchmarks\src\benchmark.cpp(2159,29): error C2672: 'utf8::utf32to8': no matching overloaded function found [D:\a\simdutf\simdutf\build\benchmarks\src\simdutf_benchmarks_benchmark.vcxproj]
```

**Comment by bkmgit** [comment]
@lemire Can you run the tests again?  The error seems to occur in other codes https://github.com/bfgroup/duck_invoke/issues/1 so rather than update compiler options have used the older interface in utfcpp, so this error no longer occurs.  A couple of tests fail on my branch when using GitHub actions, but these seem related to trying to use x86 on ARM and in utf8sse which have not changed.

Assuming tests pass, should I rebase and squash commits before asking to merge?

File: benchmarks/src/benchmark.cpp [hunk 1]
```diff
@@ -2069,7 +2069,7 @@ void Benchmark::run_convert_utf32_to_utf16_llvm(size_t iterations) {
 void Benchmark::run_convert_utf8_to_utf16_utfcpp(size_t iterations) {
     const char*  data = reinterpret_cast<const char*>(input_data.data());
     const size_t size = input_data.size();
-    std::string_view data_string = std::string_view(data,size);
+    std::string_view data_string = std::string_view(data);
```
Comment (lemire) [inline benchmarks/src/benchmark.cpp]: Why did you change `std::string_view(data,size);` into `std::string_view(data);` ?

File: benchmarks/src/benchmark.cpp [hunk 2]
```diff
@@ -2124,7 +2124,7 @@ void Benchmark::run_convert_utf16_to_utf8_utfcpp(size_t iterations) {
 void Benchmark::run_convert_utf8_to_utf32_utfcpp(size_t iterations) {
     const char*  data = reinterpret_cast<const char*>(input_data.data());
     const size_t size = input_data.size();
-    std::string data_string = std::string(data, size);
+    std::string_view data_string = std::string_view(data);
```
Comment (lemire) [inline benchmarks/src/benchmark.cpp]: Same question here... why not provide the std::string_view constructor with the size parameter? (I guess that the question applies throughout.)

**Comment by lemire** [comment]
Running tests.

File: benchmarks/src/benchmark.cpp (hunk 1 above)
Comment (bkmgit) [reply benchmarks/src/benchmark.cpp]: The C++17 and C++11 interfaces caused errors with MSVC compiler:
```
D:\a\simdutf\simdutf\benchmarks\src\benchmark.cpp(2077,30): error C2672: 'utf8::utf8to16': no matching overloaded function found [D:\a\simdutf\simdutf\build\benchmarks\src\simdutf_benchmarks_benchmark.vcxproj]
  D:\a\simdutf\simdutf\benchmarks\src\benchmark.cpp(2110,30): error C2672: 'utf8::utf16to8': no matching overloaded function found [D:\a\simdutf\simdutf\build\benchmarks\src\simdutf_benchmarks_benchmark.vcxproj]
  D:\a\simdutf\simdutf\benchmarks\src\benchmark.cpp(2131,29): error C2672: 'utf8::utf8to32': no matching overloaded function found [D:\a\simdutf\simdutf\build\benchmarks\src\simdutf_benchmarks_benchmark.vcxproj]
  D:\a\simdutf\simdutf\benchmarks\src\benchmark.cpp(2159,29): error C2672: 'utf8::utf32to8': no matching overloaded function found [D:\a\simdutf\simdutf\build\benchmarks\src\simdutf_benchmarks_benchmark.vcxproj]
  ```
  
Could not determine the reason for this, though it seems to occur elsewhere, https://github.com/bfgroup/duck_invoke/issues/1
  
Therefore removed these interfaces and used  interface described at https://github.com/nemtrif/utfcpp#utf8utf16to8-2

The error does not occur in the original utfcpp repository, but some other error does occur. Trying to get a more complete test suite added https://github.com/nemtrif/utfcpp/pull/98

Comment (lemire) [reply benchmarks/src/benchmark.cpp]: I understand, but I refer to the string_view constructor specifically. Your code might be correct, but I am just curious.

**Comment by bkmgit** [comment]
Tests seem to pass here.  The current implementation has checks for input data validity. There is also an option to convert assuming input data is valid. Can add this either in this pull request or a later one if helpful.

File: benchmarks/src/benchmark.cpp (hunk 1 above)
Comment (bkmgit) [reply benchmarks/src/benchmark.cpp]: Looked up different ways to call this to see if it might resolve the problem that `no matching overloaded function could be found`.

Comment (bkmgit) [reply benchmarks/src/benchmark.cpp]: My understanding is that the size is not a direct input for u16string_view or u32string_view:
https://github.com/cplusplus/draft/blob/main/source/strings.tex#L564
https://www.learncpp.com/cpp-tutorial/introduction-to-stdstring_view/ 

string_view does seem to have advantages, but if it is not portable for all C++17 compilers, then the preference is to use the portable version of the code. Can check again with size present in case I missed something.

**Comment by lemire** [comment]
Looks good, merging.

**Comment by lemire** [comment]
Great work.

**Comment by bkmgit** [comment]
Thanks for the review.
